### PR TITLE
add support for Netgear EAX12 / EX6250v2 series

### PR DESCRIPTION
**Pull request by looi**
### image: add additional fields to Netgear encrypted image 

These fields are used for EAX12 and EX6250v2 series, and perhaps other
devices. Compatibility is preserved with the WAX202 and WAX206.

In addition, adds the related vars to DEVICE_VARS so that the variables
work correctly with multiple devices.

References in GPL source:
https://www.downloads.netgear.com/files/GPL/EAX12_EAX11v2_EAX15v2_GPL_V1.0.3.34_src.tar.gz

* tools/imgencoder/src/gj_enc.c
  Contains code that generates the encrypted image.

### ramips: add support for Netgear EAX12 series 

Netgear EAX12, EAX11v2, EAX15v2 are wall-plug 802.11ax (Wi-Fi 6)
extenders that share the SoC, WiFi chip, and image format with the
WAX202.

Specifications: 
* MT7621, 256 MiB RAM, 128 MiB NAND 
* MT7915: 2.4/5 GHz 2x2 802.11ax (DBDC)
* Ethernet: 1 port 10/100/1000
* UART: 115200 baud (labeled on board)

All LEDs and buttons appear to work without state_default.

Installation:
* Flash the factory image through the stock web interface, or TFTP to
  the bootloader. NMRP can be used to TFTP without opening the case.

Revert to stock firmware:
* Flash the stock firmware to the bootloader using TFTP/NMRP.

References in GPL source:
https://www.downloads.netgear.com/files/GPL/EAX12_EAX11v2_EAX15v2_GPL_V1.0.3.34_src.tar.gz

* target/linux/ramips/dts/mt7621-rfb-ax-nand.dts
  DTS file for this device.

### mediatek: add support for Netgear EX6250v2 series

Netgear EX6250v2, EX6400v3, EX6410v2, EX6470 are wall-plug 802.11ac
(Wi-Fi 5) extenders. Like other MT7629 devices, Wi-Fi does not work
currently as there is no driver.

Related: https://github.com/openwrt/openwrt/pull/5084

For future reference, 2.4GHz MAC = LAN+1, 5GHz MAC = LAN+2.

Specifications: 
* MT7629, 256 MiB RAM, 16 MiB SPI NOR
* MT7761N (2.4GHz) / MT7762N (5GHz) - no driver
* Ethernet: 1 port 10/100/1000
* UART: 115200 baud (labeled on board)

Installation:
* Flash the factory image through the stock web interface, or TFTP to
  the bootloader. NMRP can be used to TFTP without opening the case.
* After installation, perform a factory reset. Wait for the device to
  boot, then hold the reset button for 10 seconds. This is needed
  because sysupgrade in the stock firmware will attempt to preserve its
  configuration using sysupgrade.tgz.
  See https://github.com/openwrt/openwrt/pull/4182

Revert to stock firmware:
* Flash the stock firmware to the bootloader using TFTP/NMRP.

### CODE REVIEW

File: target/linux/ramips/dts/mt7621_netgear_eax12.dts [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+			label = "green:wps";
+			gpios = <&gpio 17 GPIO_ACTIVE_HIGH>;
+		};
+
+		led_eth_green: eth_green {
+			label = "green:eth";
+			gpios = <&gpio 10 GPIO_ACTIVE_LOW>;
+		};
+
+		led_eth_yellow: eth_yellow {
```
Comment (musashino205) [inline target/linux/ramips/dts/mt7621_netgear_eax12.dts:82]: Please remove unused labels.

**Comment by Djfe** [comment]
Looking forward to seeing this device added :)
I'm wondering whether this just adds support for the US variant of the devices and european devices need changes to the way the image is created.

File: target/linux/ramips/dts/mt7621_netgear_eax12.dts (hunk 1 above)
Comment (looi) [reply target/linux/ramips/dts/mt7621_netgear_eax12.dts]: Done

**Comment by looi** [comment]
> Looking forward to seeing this device added :) I'm wondering whether this just adds support for the US variant of the devices and european devices need changes to the way the image is created.

I only have US devices, but I believe that all devices use the same stock firmware image and the region is controlled by the Config partition. For example, you can see on [netgear.fr](https://www.netgear.fr/support/product/eax12.aspx#download) that the linked firmware image is the same as the US model.

File: target/linux/mediatek/dts/mt7629-netgear-ex6250-v2.dts [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+			gpios = <&pio 16 GPIO_ACTIVE_LOW>;
+		};
+
+		eth_yellow {
+			label = "yellow:eth";
+			gpios = <&pio 12 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	memory@40000000 {
```
Comment (Djfe) [inline target/linux/mediatek/dts/mt7629-netgear-ex6250-v2.dts:101]: What did you define memory for? Afaik Linux does autodetection of RAM now, but I could be wrong.

File: target/linux/ramips/image/mt7621.mk [hunk 3]
```diff
@@ -1605,6 +1605,37 @@ define Device/mts_wg430223
...
+  DEVICE_ALT1_MODEL := EAX15
+  DEVICE_ALT1_VARIANT := v2
+  DEVICE_PACKAGES := kmod-mt7915-firmware -uboot-envtools
+  NETGEAR_ENC_MODEL := EAX12
+  NETGEAR_ENC_REGION := US
+  NETGEAR_ENC_HW_ID_LIST := 1010000004540000_NETGEAR
+  NETGEAR_ENC_MODEL_LIST := EAX12;EAX11v2;EAX15v2
+  BLOCKSIZE := 128k
+  PAGESIZE := 2048
+  UBINIZE_OPTS := -E 5
```
Comment (DragonBluep) [inline target/linux/ramips/image/mt7621.mk:1625]: Please replace these nand common definitions with `$(Device/nand)`.
ref: https://github.com/openwrt/openwrt/commit/b09a83808a3a4f6ff2ff0cf1f12fdc4193357896

Comment (looi) [reply target/linux/ramips/image/mt7621.mk]: Done

File: target/linux/ramips/mt7621/base-files/lib/preinit/02_sysinfo_fixup [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+do_sysinfo_ramips_fixup() {
+	local model=""
+
+	case $(board_name) in
+	netgear,eax12)
+		local part=$(find_mtd_part Config)
+		local config_model=$(strings "$part" | sed -n 's/^model=//p')
+		# EAX12, EAX11v2, EAX15v2
+		[ -n "$config_model" ] && model="Netgear $config_model"
```
Comment (hauke) [inline target/linux/ramips/mt7621/base-files/lib/preinit/02_sysinfo_fixup:11]: I prefer if we do not manually update this even when it does not show the correct name. If the hardware is anyway the same it should not make a big difference.

File: target/linux/mediatek/dts/mt7629-netgear-ex6250-v2.dts (hunk 2 above)
Comment (hauke) [reply target/linux/mediatek/dts/mt7629-netgear-ex6250-v2.dts:101]: Please remove this

File: target/linux/mediatek/mt7629/base-files/lib/preinit/02_sysinfo_fixup [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+do_sysinfo_mediatek_fixup() {
+	local model=""
+
+	case $(board_name) in
+	netgear,ex6250-v2)
+		local part=$(find_mtd_part Config)
+		local config_model=$(strings "$part" | sed -n 's/^bmodel=//p')
+		# EX6250v2, EX6400v3, EX6470, EX6410v2
+		[ -n "$config_model" ] && model="Netgear $config_model"
```
Comment (hauke) [inline target/linux/mediatek/mt7629/base-files/lib/preinit/02_sysinfo_fixup:11]: I prefer if we do not manually update this even when it does not show the correct name. If the hardware is anyway the same it should not make a big difference.

File: target/linux/mediatek/dts/mt7629-netgear-ex6250-v2.dts (hunk 2 above)
Comment (looi) [reply target/linux/mediatek/dts/mt7629-netgear-ex6250-v2.dts]: Done

File: target/linux/ramips/mt7621/base-files/lib/preinit/02_sysinfo_fixup (hunk 4 above)
Comment (looi) [reply target/linux/ramips/mt7621/base-files/lib/preinit/02_sysinfo_fixup]: Done. I also deleted the ath79 version of this file for consistency.

File: target/linux/mediatek/mt7629/base-files/lib/preinit/02_sysinfo_fixup (hunk 5 above)
Comment (looi) [reply target/linux/mediatek/mt7629/base-files/lib/preinit/02_sysinfo_fixup]: Done. I also deleted the ath79 version of this file for consistency.

**Comment by hauke** [comment]
Thanks. I applied the first 3 patches, but not the last one removing the detection in ath79. I am not sure how we want to do this in general in OpenWrt, I just didn't want to introduce more of these runtime version detection scripts. Maybe you could create a separate pull request with that patch to start a broader discussion about this topic.

**Comment by looi** [comment]
Sounds good, thanks!

**Pull request merged**